### PR TITLE
add some docs for QR objects, fixed an error in QRCompactWY

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

**Review by andreasnoack — APPROVED** [review]

File: base/linalg/qr.jl [hunk 1]
```diff
@@ -1,14 +1,50 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
 # QR and Hessenberg Factorizations
+"""
+    QR <: Factorization
+
+Represents a QR factorization of an `m`×`n`  matrix `A`, with the `Q` matrix stored as a sequence of Householder reflectors:
```
Comment (Sacha0) [inline base/linalg/qr.jl]: Should "Represents" be "Represent" for consistency with other documentation? Or does that guideline not apply to docstrings for types?

Also an extra space after the last ``` ` ``` in ``` an `m`x`n`  matrix```?

File: base/linalg/qr.jl [hunk 2]
```diff
@@ -1,14 +1,50 @@
...
+
+```math
+Q = \\prod_{i=1}^{\\min(m,n)} (I - \\tau_i v_i v_i^T).
+```
+
+The object has two fields:
 
+* `factors`: an `m`×`n` matrix of the same dimension as `A`.
+  - The upper triangular part contains ``R``, that is `triu(F.factors)` for a `QR` object `F`.
```
Comment (Sacha0) [inline base/linalg/qr.jl]: Is ``` ``R`` ``` correct?

File: base/linalg/qr.jl [hunk 3]
```diff
@@ -1,14 +1,50 @@
...
     τ::Vector{T}
     QR{T,S}(factors::AbstractMatrix{T}, τ::Vector{T}) where {T,S<:AbstractMatrix} = new(factors, τ)
 end
 QR(factors::AbstractMatrix{T}, τ::Vector{T}) where {T} = QR{T,typeof(factors)}(factors, τ)
+
 # Note. For QRCompactWY factorization without pivoting, the WY representation based method introduced in LAPACK 3.4
+"""
+    QRCompactWY <: Factorization
+
+Represents a QR factorization of an `m`×`n`  matrix `A`, with the `Q` matrix stored as a sequence of Householder reflectors in a compact format amenable to matrix multiplication operations:
```
Comment (Sacha0) [inline base/linalg/qr.jl]: Likewise here re. "Represents" and the extra space?

File: base/linalg/qr.jl (hunk 1 above)
Comment (simonbyrne) [inline base/linalg/qr.jl]: What guideline are you referring to?

**Comment by simonbyrne** [comment]
I've pushed some more changes, please feel free to make any other changes directly to the PR.

File: base/linalg/qr.jl [hunk 4]
```diff
@@ -1,21 +1,127 @@
...
+"""
+    QR <: Factorization
+
+Represents a QR factorization of an `m`×`n` matrix `A`,
+
+```math
+A = Q R
+```
+
+where ``P`` is a permutation matrix, ``Q`` is an orthogonal/unitary matrix and ``R`` is
```
Comment (tkelman) [inline base/linalg/qr.jl]: P doesn't appear anywhere else in this docstring?

File: base/linalg/qr.jl (hunk 1 above)
Comment (tkelman) [inline base/linalg/qr.jl]: probably https://github.com/JuliaLang/julia/blob/master/doc/src/manual/documentation.md

**Comment by tkelman** [comment]
what did I do here that managed to break the cross-references?

File: base/linalg/qr.jl [hunk 5]
```diff
@@ -1,21 +1,122 @@
...
+The matrix `Q` is stored as a sequence of Householder reflectors ``v_i``
+and coefficients ``\\tau_i`` where:
+
+```math
+Q = \\prod_{i=1}^{\\min(m,n)} (I - \\tau_i v_i v_i^T).
+```
+
+The object has two fields:
 
+* `factors` is an `m`×`n` matrix of the same dimension as `A`.
```
Comment (Sacha0) [inline base/linalg/qr.jl]: Perhaps "dimension" should be "dimensions" or "size"?

File: base/linalg/qr.jl [hunk 6]
```diff
@@ -1,21 +1,122 @@
...
+
+A QR matrix factorization stored in a packed format, typically obtained from
+[`qrfact`](@ref). If `A` is an `m`×`n` matrix, then
+
+```math
+A = Q R
+```
+
+where ``Q`` is an orthogonal/unitary matrix and ``R`` is upper triangular.
+The matrix `Q` is stored as a sequence of Householder reflectors ``v_i``
```
Comment (Sacha0) [inline base/linalg/qr.jl]: For consistency with the preceding sentence, perhaps ``` ``Q`` ```?

File: base/linalg/qr.jl [hunk 7]
```diff
@@ -1,21 +1,122 @@
...
 end
 QR(factors::AbstractMatrix{T}, τ::Vector{T}) where {T} = QR{T,typeof(factors)}(factors, τ)
+
 # Note. For QRCompactWY factorization without pivoting, the WY representation based method introduced in LAPACK 3.4
+"""
+    QRCompactWY <: Factorization
+
+A QR matrix factorization stored in a compact blocked format, typically obtained from
+[`qrfact`](@ref). It is similar to a [`QR`](@ref) object except that the
+orthogonal/unitary matrix `Q` is stored in *Compact WY* format [^Schreiber1989], as a
```
Comment (Sacha0) [inline base/linalg/qr.jl]: Single/double quote consistency about `Q`?

File: base/linalg/qr.jl [hunk 8]
```diff
@@ -1,21 +1,122 @@
...
+Q = \\prod_{i=1}^{\\min(m,n)} (I - \\tau_i v_i v_i^T) = I - V T V^T
+```
+
+such that ``v_i`` is the ``i``th column of ``V``, and ``\tau_i`` is the ``i``th diagonal
+element of ``T``.
+
+The object has two fields:
+
+* `factors`, as in the [`QR`](@ref) type, is an `m`×`n` matrix of the same
+  dimension as `A`.
```
Comment (Sacha0) [inline base/linalg/qr.jl]: "dimension" vs "dimensions" or "size" here as well?

**Comment by simonbyrne** [comment]
@Sacha0 I think I got them all. Please feel free to edit directly if you see anything else.

File: base/linalg/qr.jl [hunk 9]
```diff
@@ -57,8 +57,7 @@ element of ``T``.
 
 The object has two fields:
 
-* `factors`, as in the [`QR`](@ref) type, is an `m`×`n` matrix of the same
-  dimension as `A`.
+* `factors`, as in the [`QR`](@ref) type, is an `m`×`n` matrix.
```
Comment (tkelman) [inline base/linalg/qr.jl]: m and n aren't introduced earlier in this one's docstring

File: base/linalg/qr.jl [hunk 10]
```diff
@@ -343,13 +426,25 @@ end
...
+"""
 struct QRPackedQ{T,S<:AbstractMatrix} <: AbstractMatrix{T}
     factors::S
     τ::Vector{T}
     QRPackedQ{T,S}(factors::AbstractMatrix{T}, τ::Vector{T}) where {T,S<:AbstractMatrix} = new(factors, τ)
 end
 QRPackedQ(factors::AbstractMatrix{T}, τ::Vector{T}) where {T} = QRPackedQ{T,typeof(factors)}(factors, τ)
 
+"""
+    QRPackedQ <: AbstractMatrix
```
Comment (tkelman) [inline base/linalg/qr.jl]: wrong name here

**Review by Sacha0 — APPROVED** [review]
These additions are fantastic @simonbyrne! :)

**Pull request merged**